### PR TITLE
Add support for limited Announce Activity

### PR DESCRIPTION
**Pull request by noellabo**
For Announce Activity, we'll add support for Bearcaps dereference, limited visibility, and direct delivery of actors inbox.

### CODE REVIEW

**Review by Gargron — APPROVED** [review]

File: app/lib/activitypub/activity/announce.rb [hunk 1]
```diff
@@ -34,6 +29,84 @@ def perform
...
+    @params[:visibility] = :limited if @params[:visibility] == :direct
+  end
+
+  def postprocess_audience_and_deliver
+    return if @status.mentions.find_by(account_id: @options[:delivered_to_account_id])
+
+    delivered_to_account = Account.find(@options[:delivered_to_account_id])
+
+    @status.mentions.create(account: delivered_to_account, silent: true)
+    @status.update(visibility: :limited) if @status.direct_visibility?
```
Comment (ClearlyClaire) [inline app/lib/activitypub/activity/announce.rb:96]: Not a big deal, but not useful here as reblogs can't have `direct` visibility.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
While the feature would be somewhat useful, this current implementation has several performance issues.

Furthermore, I'm uncomfortable with the whole audience processing logic being duplicated with `Create`.

Finally, I am not too sure mentions for reblogs are handled in a way that makes sense across the codebase.

File: app/lib/activitypub/activity/announce.rb (hunk 1 above)
Comment (noellabo) [reply app/lib/activitypub/activity/announce.rb]: Moved duplicate code for Create and Announce to one place.

File: app/lib/activitypub/activity/announce.rb [hunk 2]
```diff
@@ -2,28 +2,23 @@
 
 class ActivityPub::Activity::Announce < ActivityPub::Activity
   def perform
+    dereference_object!
```
[the remark this answers is not in the document]
Comment (noellabo) [reply app/lib/activitypub/activity/announce.rb]: Changed to check if url points to local before `reblog_of_local_status?` accesses the database. After passing these checks, I now run `dereference_object!`.

File: app/lib/activitypub/activity/announce.rb [hunk 3]
```diff
@@ -2,28 +2,23 @@
...
+
     return reject_payload! if delete_arrived_first?(@json['id']) || !related_to_local_activity?
 
     RedisLock.acquire(lock_options) do |lock|
       if lock.acquired?
-        original_status = status_from_object
-
-        return reject_payload! if original_status.nil? || !announceable?(original_status)
+        @original_status = status_from_object
```
[the remark this answers is not in the document]
Comment (noellabo) [reply app/lib/activitypub/activity/announce.rb:10]: Changed to record the success or failure of dereference and use the acquired object if the dereference is successful.

**Comment by noellabo** [comment]
This PR allows you to receive boosts for circle members in the circle function. The ability to deliver limited annouce will be implemented separately.

We have also made a change to allow dereferenced Announce Activities delivered by group actors.
https://github.com/tootsuite/mastodon/blob/23f219a1966485b4417ef42c4f7901bb45b2db25/app/lib/activitypub/activity/announce.rb#L66-L68

This allows the group server to deliver posts that are exclusive to group participants.

I haven't tested it enough yet, so I'll verify it over the weekend.

**Comment by noellabo** [comment]
Changed to use the object fetched by dereference and solved the performance problem.

The code related to the group will be reverted and will be handled by another PR.

Please review.

**Review by ClearlyClaire — COMMENTED** [review]
Not marked as accepted as I would like another set of eyes on this PR that might have subtle security implications.

I have not found anything fishy, though, and the change seems good to me.

I'm not too sure about `process_audience` and `postprocess_audience_and_deliver` being in `ActivityPub::Activity` because they rely on rely specific `params`, `status` and `mentions` instance variables and are really specific to status creation, but I still think it's better than duplicating them. Maybe a concern?

**Comment by noellabo** [comment]
I moved the methods related to process_audience to ProcessAudience concern.

**Review by ClearlyClaire — COMMENTED** [review]
Again, not marked as accepted as I would really like another pair of eyes checking this PR for security implications, but it looks good to me and I have no further remarks at the moment!

**Comment by noellabo** [comment]
The evaluation has not progressed, but I think that it will be reviewed when the group function is used. Stay open.

**Comment by noellabo** [comment]
This pull request will remain closed and will be pulled again when the need arises.

**Pull request closed without merging**